caikit/runtime/service_generation/protoable.py

```diff
@@ -0,0 +1,158 @@
+            # if there's only 1 protoable found, return that
+            if len(union_protoables) == 1:
+                return union_protoables[0]
+            # otherwise, try to get the data model objects in the Union
+            dm_types = [
+                arg
+                for arg in union_protoables
+                if inspect.isclass(arg) and issubclass(arg, DataBase)
+            ]
+            # if there are multiple, pick the first one
+            if len(dm_types) > 0:
+                log.debug2(
+                    "Picking first data model type %s in union protoables %s",
+                    dm_types,
+                    union_protoables,
+                )
+                return dm_types[0]
+            log.debug(
+                "Just picking first protoable type %s in union",
+                union_protoables[0],
+            )
+            return union_protoables[0]
+        return arg_type
+    log.debug("Skipping non-protoable argument type [%s]", arg_type)
+
+
+def extract_data_model_type_from_union(arg_type: Type) -> Type:
+    """Helper function that determines the right data model type to use from a Union"""
+
+    # Decompose this type using typing to determine if it's a useful typing hint
+    typing_origin = get_origin(arg_type)
+    typing_args = get_args(arg_type)
+
+    # If this is a data model type, no need to do anything
+    if isinstance(arg_type, type) and issubclass(arg_type, DataBase):
+        return arg_type
+
+    # Handle Unions by looking for a data model object in the union
+    if typing_origin is Union:
+        dm_types = [
+            arg
+            for arg in typing_args
+            if inspect.isclass(arg) and issubclass(arg, DataBase)
+        ]
+        if dm_types:
+            log.debug2(
+                "Found data model types in Union: [%s], taking first one", dm_types
+            )
+            return extract_data_model_type_from_union(dm_types[0])
+
+    # if it's anything else we just return as is
+    # we don't actually want to throw errors from service generation
+    log.warning("Return type [%s] not a DM type, returning as is", arg_type)
+    return arg_type
+
+
+def is_protoable_type(arg_type: Type) -> bool:
+    """
+    Returns True is arg_type is in PROTO_TYPE_MAP(float, int, bool, str, bytes)
+    Or if it's an imported Caikit data model class.
+    Or if it's a Union of at least one of those.
+    Or if it's a List of one of those.
+    False otherwise"""
+    proto_primitive_set = list(PROTO_TYPE_MAP.keys())
+
+    if arg_type in proto_primitive_set:
+        return True
+    if isinstance(arg_type, type) and issubclass(arg_type, DataBase):
+        return True
+
+    if typing.get_origin(arg_type) == list:
+        log.debug2("Arg is List")
+        # check that list is not nested
+        if len(typing.get_args(arg_type)) == 1:
+            return typing.get_args(arg_type)[0] in proto_primitive_set
+        # TODO: that check is always true
```

Comment:
which check is always true? This is a bit unclear - could this be rephrased to what needs to be done?

Reply:
I think the `List[]` type hint always takes exactly one arg. It can do no args, though, I think

Reply:
Looks like it can only do a single arg, zero and >1 args both fail

```
>>> foo: List[]
  File "<stdin>", line 1
    foo: List[]
              ^
SyntaxError: invalid syntax
>>> foo: List[int, float]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/homebrew/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/typing.py", line 351, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/typing.py", line 1556, in __getitem__
    _check_generic(self, params, self._nparams)
  File "/opt/homebrew/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/typing.py", line 281, in _check_generic
    raise TypeError(f"Too {'many' if alen > elen else 'few'} arguments for {cls};"
TypeError: Too many arguments for typing.List; actual 2, expected 1
```

Reply:
and for the simple `list` case:
```
>>> typing.get_origin(list) == None
True
```

Reply:
```py
>>> from typing import List, get_origin
>>> get_origin(List)
<class 'list'>
```

Reply:
So I think you need to keep the check rather than moving to an `assert` (which is what I was going to suggest)